update I420 buffer handling and stride calculations

JuunKR opened the pull request:
# Refactor I420 buffer handling and stride calculations (odd resolution fix)

### Background

While using `node-webrtc` for WebRTC testing on an iPhone device, I encountered the following error:

```
TypeError: Expected a .byteLength of 387477, not 387476
    at VideoProcessor.processFrame (/app/src/service/VideoProcessor.ts:122:13)
    at /app/node_modules/@roamhq/wrtc/lib/eventtarget.js:43:18
    at Set.forEach (<anonymous>)
    at /app/node_modules/@roamhq/wrtc/lib/eventtarget.js:36:15
    at processTicksAndRejections (node:internal/process/task_queues:85:11)
[ERROR] 02:02:52 TypeError: Expected a .byteLength of 387477, not 387476
```

During debugging, I found that in some cases the frame dimensions included **odd values** such as `width = 678` and `height = 381`.
The current `node-webrtc` implementation did not properly account for odd resolutions in I420 buffer calculations.

---

### Changes

* **Odd resolution handling**

  * Updated I420 buffer size calculation to round up remainders when dimensions are odd.
  * Referenced WebKit’s libwebrtc implementation, which already includes proper handling:

    * [WebKit i010\_buffer.cc](https://github.com/WebKit/WebKit/blob/main/Source/ThirdParty/libwebrtc/Source/webrtc/api/video/i010_buffer.cc)

* **Stride calculation improvement**

  * Replaced manual branching logic with a libyuv function that provides stride calculation.
  * Since libyuv is already used elsewhere in the codebase, this improves consistency and maintainability.

    * [[libyuv convert.cc](https://github.com/WebKit/WebKit/blob/main/Source/ThirdParty/libwebrtc/Source/third_party/libyuv/source/convert.cc)](https://github.com/WebKit/WebKit/blob/main/Source/ThirdParty/libwebrtc/Source/third_party/libyuv/source/convert.cc)

---

### Expected Outcome

* Fixes `.byteLength mismatch` errors when processing frames with odd resolutions.
* Simplifies stride calculation logic and ensures consistency by leveraging libyuv functions.

---

### Notes

This is my **first open source contribution** 
If I missed any procedural steps or if there are improvements I should make in the PR process, please let me know.

JuunKR committed:
update I420 buffer handling and stride calculations

- Updated sizeOfChromaPlane and strideU methods in image_data.hh to correctly handle odd resolutions and use proper calculations based on image dimensions.
- Refactored video_frame_buffer.cc to utilize libyuv for I420 buffer copying, improving performance and reliability.
- Adjusted expected byte length calculations in I420Helpers to align with the new chroma plane size logic, adding proper handling for odd resolutions.

JuunKR commented:
Just to add: I built and tested this change locally, and confirmed that it works as expected.

hazelduvall commented on src/dictionaries/node_webrtc/image_data.hh:
```diff
@@ -45,7 +45,7 @@ public:
   uint8_t *dataU() { return &dataY()[sizeOfLuminancePlane()]; }
 
-  [[nodiscard]] int strideU() const { return width() / 2; }
+  [[nodiscard]] int strideU() const { return (width() + 1) / 2; }
```
This change aligns `strideU()` and `strideV()` with what's used in libwebrtc's I420Buffer, nice:

https://github.com/WebKit/WebKit/blob/36cb62dfa5e000484810987936bee8ff548fe9fa/Source/ThirdParty/libwebrtc/Source/webrtc/api/video/i420_buffer.cc#L51-L52

hazelduvall commented on src/dictionaries/node_webrtc/image_data.hh:
```diff
@@ -36,7 +36,7 @@ public:
 
   [[nodiscard]] size_t sizeOfChromaPlane() const {
-    return sizeOfLuminancePlane() / 4;
+    return static_cast<size_t>((width() + 1) / 2) * static_cast<size_t>((height() + 1) / 2);
```
This change aligns the `dataV()` function with libwebrtc's I420Buffer, nice:

https://github.com/WebKit/WebKit/blob/36cb62dfa5e000484810987936bee8ff548fe9fa/Source/ThirdParty/libwebrtc/Source/webrtc/api/video/i420_buffer.cc#L154-L162

hazelduvall commented on src/dictionaries/webrtc/video_frame_buffer.cc:
```diff
@@ -40,18 +41,19 @@ TO_NAPI_IMPL(const webrtc::I420BufferInterface *, pair) {
-  auto sizeOfDstYPlane = value->width() * value->height();
-  auto sizeOfDstUPlane = sizeOfDstYPlane / 4;
-  auto sizeOfDstVPlane = sizeOfDstYPlane / 4;
+  auto dstYStride  = value->width();
+  auto dstUVStride = (value->width() + 1) / 2;
+  auto uvHeight     = (value->height() + 1) / 2;
 
+  auto sizeOfDstYPlane = dstYStride  * value->height();
+  auto sizeOfDstUPlane = dstUVStride * uvHeight;
+  auto sizeOfDstVPlane = dstUVStride * uvHeight;
```
I420Buffer calculates these (and by extension, `dst{Y,U,V}Plane`) using the `value->Stride{Y,U,V}()` and `value->MutableData{Y,U,V}()` functions. I think we should use those instead here (since they will result in the same values, and it's conceptually a bit simpler):

https://github.com/WebKit/WebKit/blob/36cb62dfa5e000484810987936bee8ff548fe9fa/Source/ThirdParty/libwebrtc/Source/webrtc/api/video/i420_buffer.cc#L89-L112

hazelduvall commented on src/methods/i420_helpers.cc:
```diff
@@ -25,8 +25,13 @@ Validation<RgbaImageData> ImageData::toRgba() const {
+  auto v = (imageData.width + 1) / 2;
+  auto h = imageData.height;
+
+  size_t expectedByteLength = static_cast<size_t>(y * h + (u + v) * ((h + 1) / 2));
```
This is also the function that I420Buffer seems to use, nice:

https://github.com/WebKit/WebKit/blob/36cb62dfa5e000484810987936bee8ff548fe9fa/Source/ThirdParty/libwebrtc/Source/webrtc/api/video/i420_buffer.cc#L46

hazelduvall reviewed (commented):
Thanks for the PR!! This is a big improvement, thanks for taking it on. Just one outstanding comment besides the unresolved one below: can we add a testcase to `test/i420helpers.js` that ensures this new functionality is handled correctly? Probably using your image dimensions that turned out to be bad.

JuunKR commented:
I’ll make the changes today and ensure the build and tests go well

JuunKR committed:
Add tests for odd-sized I420 to RGBA and RGBA to I420 conversions

- Implemented new test cases for converting I420 frames to RGBA and vice versa, specifically for odd dimensions.
- Updated sizeOfChromaPlane calculation in I420Frame class to ensure accurate chroma plane size for odd resolutions.

JuunKR commented:
Thanks for the suggestion — I also checked the code in I420Buffer::Copy.
In our case, the destination buffer is expected to be tightly packed (Y stride = width, UV stride = ceil(width/2)), and our byteLength and helper/tests rely on that assumption (W*H + 2*ceil(W/2)*ceil(H/2)).
If we switch the destination to value->Stride{Y,U,V}(), the plane sizes become stride-based (potentially padded), which would be larger than our expected packed size and cause mismatches — especially for odd dimensions.

I might have misunderstood your request, so please double-check and let me know if I’m missing something.

Also, I added test cases for odd dimensions (678×381) by reusing the existing ones, since the original bug occurred with odd sizes. I also fixed the UV calculation in the test code to handle odd sizes correctly.

hazelduvall commented:
Ah, I see your concern: `value` is of type `webrtc::I420BufferInterface*`, which could be non-compact, unlike our `I420ImageData` representation. We want to compact into `data`, which is just a `uint8_t*` and has no associated stride information, so we create some here.

In that case, I think we're all good to go! Sorry about my confusion and thanks for the explanation. I'll merge this and cut a release once I remove some debug code from the `develop` branch. Thanks again for the PR!!

sebmellen commented:
Hey @duvallj — do you have any ETA on when this will be packaged up for use?

This would be a massive unblock for me and @SarjuHansaliya, as we're relying on this library for a core internal product.

@JuunKR thank you for the PR!

hazelduvall commented:
Hi @sebmellen ! I've just made a v0.9.1 release of `@roamhq/wrtc` that includes this change. Let me know if you have any troubles with it!

SarjuHansaliya commented:
Thank you so much @duvallj appreciated 🤝
